### PR TITLE
Add hexo-filter-image plugin

### PR DESCRIPTION
**Pull request by JoeyBling**
## Check List

**Please read and check followings before submitting a PR.**

- [ ] I want to publish my theme on Hexo official website.
    - [ ] I have read the [theme publishing doc](https://hexo.io/docs/themes#Publishing).
    - [ ] `link` URL is correct.
    - [ ] `preview` URL is correct.
    - [ ] `preview` URL web site is rendered correctly.
    - [ ] Add a screenshot to `source/themes/screenshots`.
    - [ ] Screenshot filename is same as value of `name`.
    - [ ] Screenshot size is `800 * 500`.
    - [ ] Screenshot file format is `png`.
- [x] I want to publish my plugin on Hexo official website.
    - [x] I have read the [plugin publishing doc](https://hexo.io/docs/plugins#Publishing).
    - [x] `link` URL is correct.
- [ ] Others (Update, fix, translation, etc...)

### CODE REVIEW

**Comment by curbengh** [comment]
Please try not to use the master branch (of your fork) for pull request, otherwise you'll end up with the same issue as https://github.com/hexojs/site/pull/1069 in future PRs.

In fact, do not use master branch for any pull request even in other repos.

---

I would advice not to touch the master branch (of your fork) except to sync with upstream. To sync,
``` sh
$ git clone https://github.com/JoeyBling/site hexo-site
$ cd hexo-site
$ git remote add upstream https://github.com/hexojs/site
$ git checkout master
$ git fetch upstream
$ git rebase upstream/master
$ git push -u
```

Always push commit(s) (of your work) to a new branch. Once the pull request has been merged in the upstream, remove the branch in your local copy
```
$ git checkout master
$ git b
```

**Comment by curbengh** [comment]
Anyway, looks like this plugin is trying to resolve similar issue as [hexo-typora-image](https://github.com/deptno/hexo-typora-plugins/tree/master/packages/hexo-typora-image)? https://github.com/hexojs/site/pull/735

**Comment by JoeyBling** [comment]
> Anyway, my understanding of the purpose of this plugin is to prepend `root` value to the image path. This also can resolved by the [`img`](https://hexo.io/docs/tag-plugins#Image) tag plugin.
> 
> ```
> {% img /path/to/image %}
> ```
> 
> Perhaps hexo-renderer-marked can add a workaround for that issue.

Yes, you're right, I just chose to rewrite the plugin without breaking the `Markdown` syntax

**Comment by JoeyBling** [comment]
@curbengh Well, can't I select master branch commit request?

**Comment by curbengh** [comment]
> Yes, you're right, I just chose to rewrite the plugin without breaking the `Markdown` syntax

That's fine, I do understand some may prefer pure Markdown syntax. https://github.com/hexojs/site/pull/735 also addresses image path issue, but looks like this plugin fixes different issue.

Anyway, thanks for raising up this issue, I'm working on to fix it in [hexo-renderer-marked](https://github.com/hexojs/hexo-renderer-marked/pull/111).

> Well, can't I select master branch commit request?

you could, it's just _discouraged_, I suggest a branch so that the issue in https://github.com/hexojs/site/pull/1069 doesn't come back again (in future PRs). If you're _sure_ that this is the only or the last PR for this repo, it's fine to use the `master` branch.

**Comment by JoeyBling** [comment]
@curbengh Yeah. It's only right to choose the right plugin for you, depending on the situation

**Comment by JoeyBling** [comment]
@curbengh that's great,thanks,that is the last PR for this repo, it's can to use the master branch.

File: source/_data/plugins.yml [hunk 1]
```diff
@@ -2277,3 +2277,14 @@
...
     - customization
+- name: hexo-filter-image
+  description: 自动为hexo中的图片映射绝对路径 | Automatically maps absolute paths to images in hexo
+  link: https://github.com/JoeyBling/hexo-filter-image
+  tags:
+    - filter
+    - image
+    - path
+    - asset
+    - options
```
Comment (curbengh) [inline source/_data/plugins.yml]: please remove this tag and the rest, the plugin only have loggin option.

**Comment by curbengh** [comment]
I just tested your plugin and it works for `root: /testPath` and `root: /testPath/` 👍 .
Aside from the tag removal, can you update your readme to show `root: /testPath/`, instead of `root: /testPath`. I'm aware both works, but having a trailing slash is a preferred convention.

**Comment by JoeyBling** [comment]
> I just tested your plugin and it works for `root: /testPath` and `root: /testPath/` 👍 .
> Aside from the tag removal, can you update your readme to show `root: /testPath/`, instead of `root: /testPath`. I'm aware both works, but having a trailing slash is a preferred convention.

> Thank you for your feedback，And I solved this problem in version 1.2.1

Update dependency package use `npm update hexo-filter-image -save`
Then execute `hexo clean` to test it

**Comment by JoeyBling** [comment]
all right,I don't need to do anything else.
**finish this PR**

File: source/_data/plugins.yml (hunk 1 above)
Comment (curbengh) [reply source/_data/plugins.yml]: I meant remove `options`, `parameters` and `customization` tags.

**Comment by curbengh** [comment]
> Thank you for your feedback，And I solved this problem in version 1.2.1
> 
> Update dependency package use `npm update hexo-filter-image -save`
> Then execute `hexo clean` to test it

I meant the readme of your plugin, specifically the [Example section](https://github.com/JoeyBling/hexo-filter-image#example). Please change
`root: /testPath`
to
`root: /testPath/`

**Comment by JoeyBling** [comment]
Well, Are you didn't delete the tags for me?
so,how can i do it.create new PR?

**Comment by JoeyBling** [comment]
> > Thank you for your feedback，And I solved this problem in version 1.2.1
> > Update dependency package use `npm update hexo-filter-image -save`
> > Then execute `hexo clean` to test it
> 
> I meant the readme of your plugin, specifically the [Example section](https://github.com/JoeyBling/hexo-filter-image#example). Please change
> `root: /testPath`
> to
> `root: /testPath/`

Well, maybe I didn't understand it well enough and it has been fix.
I was wondering if the config `root: /testPath` will affect the hexo build?

File: source/_data/plugins.yml (hunk 1 above)
Comment (JoeyBling) [reply source/_data/plugins.yml]: > I meant remove `options`, `parameters` and `customization` tags.

I see you deleted it

Comment (JoeyBling) [reply source/_data/plugins.yml]: Well, I think it should be all right now.
Sorry, I rarely use it PR

Comment (curbengh) [reply source/_data/plugins.yml]: Please remove `options` tag as well.

**Comment by curbengh** [comment]
> Well, maybe I didn't understand it well enough and it has been fix.
> I was wondering if the config `root: /testPath` will affect the hexo build?

Hexo appends a trailing slash if missing, but we recommend user to always put [a trailing slash](https://github.com/hexojs/hexo-starter/blob/571320ba41a83e065d7560e050eb3fa63ad74a57/_config.yml#L15). I suggest to add a trailing slash to your readme to avoid user confusion and be consistent to the [hexo's default](https://github.com/hexojs/hexo-starter/blob/571320ba41a83e065d7560e050eb3fa63ad74a57/_config.yml#L15).

File: source/_data/plugins.yml (hunk 1 above)
Comment (JoeyBling) [reply source/_data/plugins.yml]: > Please remove `options` tag as well.

no problem,i already remove that's tag

**Comment by JoeyBling** [comment]
> > Well, maybe I didn't understand it well enough and it has been fix.
> > I was wondering if the config `root: /testPath` will affect the hexo build?
> 
> Hexo appends a trailing slash if missing, but we recommend user to always put [a trailing slash](https://github.com/hexojs/hexo-starter/blob/571320ba41a83e065d7560e050eb3fa63ad74a57/_config.yml#L15). I suggest to add a trailing slash to your readme to avoid user confusion and be consistent to the [hexo's default](https://github.com/hexojs/hexo-starter/blob/571320ba41a83e065d7560e050eb3fa63ad74a57/_config.yml#L15).

yeah,I add a trailing slash to readme,good job

**Comment by JoeyBling** [comment]
Why haven't I seen the update on the official website? Can't find my plugin.
Isn't this PR request approved now?

**Comment by JoeyBling** [comment]
Thanks.

**Pull request merged**